Makefile

```diff
@@ -2,6 +2,9 @@ CARGO=RUSTFLAGS='-F warnings' cargo
 
 .PHONY: debug release test test-release bench fmt cov clean clippy security_audit
 
+citaship:
+	make -f ./scripts/Makefile
+
```

Comment:
In your branch situation, other commands related with `cargo` in this makefile , such as `test`, `bench`  also need to be changed.

Reply:
Okay, this commit terms to solve the starters' self compilation.

Compare with the code developers who need `test`, `bench` directly, once they test the code of the whole project, use `cargo` might be more in line with their habits than `make`?_

__I'm not sure about it now__, because, in the same situation, not every developer needs to compile the code by him/herself,  so I wrote theme like above...

+ Completely separated docker operation and local operation.
+ If don't want to operate the docker container, `make test`, `make bench`, are still work.
+ If developer want to `make test`, `make bench` in the docker container, `./scripts/env.sh make test`, `./scripts/env.sh make bench` will help.
+ Compile release code, just `make` in the source dir, this might be the most commonly used.

Reply:
I found it is building very slowly in a docker container. It takes me about 1h to build it from the beginning.

Reply:
> I found it is building very slowly in a docker container. It takes me about 1h to build it from the beginning.

Yep, 40+mins once, my mac blowed up last night when I built docker container、Docerfile、release version together...

The best why to install might be using the __compiled `code`__ or a __complete `docker image`__?

Reply:
For a newcomer, he/she may prefer to `run` CITA first, rather than compiling it from source, or even learning the code.
So, If user `install` CITA, I think is better to use the compiled code or a complete docker image (I think a complete docker image is better). And,
If user `clone` CITA, that means he/she is ready to hack the code, so we need to help him/her `compile & run` from `code`.
What do you think about it?

Reply:
You mean `rm docker container` While building in scripts can get a better performance?

Reply:
yeah, the first choice should be friend for uses

Reply:
yep, I'm agree with your opinion, so:

|Mode|Description|needs|
|-|-|-|
|Run|Compiled code, with a `cita` command can run in `$PATH`|brew/snap|
|Install|Just pull a `docker image`, a `env.sh` is all they need.|Complete Docker Image|
|Clone|`env.sh make release` doesn't a complex way, they're ready for it.|keep|

1. plz check my formula~ I have just optimized it last night, it's quite nice~ 
2. If the formula is ok, I'm trying to write a Docker Image now.
3. For `make`, I'll keep it as it is ~

Reply:
nice job!

Reply:
I got some errors here:
```\$ brew tap clearloop/hyperfox
Updating Homebrew...
==> Tapping clearloop/hyperfox
Cloning into '/usr/local/Homebrew/Library/Taps/clearloop/homebrew-hyperfox'...
remote: Repository not found.
fatal: repository 'https://github.com/clearloop/homebrew-hyperfox/' not found
Error: Failure while executing; `git clone https://github.com/clearloop/homebrew-hyperfox /usr/local/Homebrew/Library/Taps/clearloop/homebrew-hyperfox --depth=1` exited with 128.
```

Reply:
oops, I wrote the wrong url, 

__try this: `tap clearloop/homebrew-hypercask`__

Reply:
Wait a minute, something occurs...I'll fix it soon XD

Reply:
Thanks, it works!
If we have another two features in `cita` script, that would be more perfect:
- Generating chain configuration, just likes:
```
./env.sh ./scripts/create_cita_config.py create --super_admin "0x4b5ae4567ad5d9fb92bc9afd6a657e6fa13a2523"  --chain_name test2-chain --jsonrpc_port 2337 --ws_port 5337 --grpc_port 6000 --nodes "127.0.0.1:8000,127.0.0.1:8001,127.0.0.1:8002,127.0.0.1:8003"
```
- Exposing a port from the running docker which likes `-p 1337:1337`.

After that, we can get a `cita-block-chain` by using `cita` which just installed.

Reply:
And then, the user can use a client (sdk, cita-cli, etc.) to access the `block-chain`, do anything they want (check the status of the chain, deploy a contract on to or something else).

Reply:
> If we have another two features in `cita` script, that would be more perfect:
> 
> * Generating chain configuration, just likes:
> 
> ```
> ./env.sh ./scripts/create_cita_config.py create --super_admin "0x4b5ae4567ad5d9fb92bc9afd6a657e6fa13a2523"  --chain_name test2-chain --jsonrpc_port 2337 --ws_port 5337 --grpc_port 6000 --nodes "127.0.0.1:8000,127.0.0.1:8001,127.0.0.1:8002,127.0.0.1:8003"
> ```
> * Exposing a port from the running docker which likes `-p 1337:1337`.
> 
> After that, we can get a `cita-block-chain` by using `cita` which just installed.

## How about these?

__For the first feature:__

`cita create` has these arguments:

|Arguments|Description|
|-|-|
|authorities|Authorities (addresses) list|
|chain_name|Name of the new chain|
|nodes|Node network addresses for new nodes|
|super_admin|Address of super admin||
|contract_arguments|Update constructor arguments for system contract.|
|time_stamp|Specify a timestamp to use|
|grpc_port|grrpc port for this chain|
|jsonrpc_port|jsonrpc port for this chain|
|ws_port|websocket port for this chain|
|enable_tls|The data is encrypted and transmitted on the network|

_Arguments might too long that not friendly for the command-line._

I have an idea, once we run `cita`, we create a `cita-block-chain` with preset arguments by default, and, the `create` command will interact with users line by line: 
```shell
$ cita create "(super_admin)"
# is super_admin the only required argument?
$ cita create -v "(super_admin)"
# the verbose tag `-v`.
please input your configure arguments, press `enter` will set default:
$ chain name: 
# after input, a new line comes or replaces the upper line's view...
$ json_rpc_port: 
...
```

I think it is no good to rewrite this method in `shell` from `python` now? Should I call the `create_cita_chain.py` from `./cita` directly? the better might be write a rust commander directly? XD

__The second feature:__

|Command|Execution|
|-|-|
|`cita` | run docker container as `-p 1337:1337`|
|`cita port 6666:6666`|run docker container as `-p 6666:6666`|

Reply:
- For the first feature, I think is ok to call `create_cita_chain.py` from `./cita` directly.
- And the second feature, your proposal looks good!

Reply:
> the better way might be writing a rust commander directly

yeah, That's the plain. #221 We can discuss this on the issue.

>  Should I call the create_cita_chain.py from ./cita directly

About this, I'd like to seperate the create_cita_config as an independent tool. The reason is that the tool should be just used to create genesis, the genesis is actually what cita needs.
I think cita does not need to consider how genesis is created. so we can have py version, Rust version, something like that.

Reply:
Independent tool sounds good!

So we also need  `brew install cita-config`(or some command like that)  to help newcomer running CITA easily.

Reply:
I've written the `port` command just now, 

A new bug occurs, and I wondering why:
```
2019-03-21T11:35:25.543589700+08:00 network connect b93f89efcac12d3f8cb2002e55b1b7df26173aaf968115fd7e376b1021187506 (container=778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca, name=bridge, type=bridge)
2019-03-21T11:35:26.017352200+08:00 container start 778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca (image=cita/cita-run:ubuntu-18.04-20181009, maintainer=Cryptape Technologies <contact@cryptape.com>, name=cita)
2019-03-21T11:35:26.197330600+08:00 container die 778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca (exitCode=0, image=cita/cita-run:ubuntu-18.04-20181009, maintainer=Cryptape Technologies <contact@cryptape.com>, name=cita)
2019-03-21T11:35:26.475652000+08:00 network disconnect b93f89efcac12d3f8cb2002e55b1b7df26173aaf968115fd7e376b1021187506 (container=778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca, name=bridge, type=bridge)
```

Here is my config code:
```
# Run container                                                                                                                                                                                             
docker ps | grep ${CONTAINER_NAME} > /dev/null 2>&1
if [ $? -ne 0 ]; then
    echo "Start docker container ${CONTAINER_NAME} ..."
    docker rm ${CONTAINER_NAME} > /dev/null 2>&1
    docker run -d \
           --net=${SYSTEM_NET} \
           --volume ${SOURCE_DIR}:${WORKDIR} \
           --volume ${LOCALTIME_PATH}:/etc/localtime \
           --volume ${HOME}/.docker_cargo/git:${CARGO_HOME}/git \
           --volume ${HOME}/.docker_cargo/registry:${CARGO_HOME}/registry \
           --env USER_ID=${USER_ID} \
           --workdir ${WORKDIR} \
           --name ${CONTAINER_NAME} \
           -p $EXPOSE ${DOCKER_IMAGE} \
           /bin/bash -c "${INIE_CMD}"
    # Wait entrypoint.sh to finish                                                                                                                                                                          
    sleep 3
fi
```

__About the `brew install cita`:__
Should just install cita compiled code without docker?

Docker environment is more like production, for company, developers just debug their code and learning cita on it.

Docker image is more than 2GB, but __compiled code zips__  are less than 50MB, it's a favor for newcomers who just want to try it.

__Sperate `cita-config`:__

It's really nice, I'll try it after fixed the bug above, 

btw, I'll also inject the create code into `cita` as just like `create_cita_config.py` does, the line-by-line interactive, if we got `cita-config`, it becomes verbose.

Reply:
Without docker, means that we need to release compiled code for different OS. There are a lot of work for test!

Reply:
Could we discuss it on your issue? so someone else can join it, too.

Reply:
> Could we discuss it on your issue? so someone else can join it, too.

Of course, the `Makefile` has already outdated, it wouldn't be sad...XD

> Without docker, means that we need to release compiled code for different OS. There are a lot of work for test!

It is a problem, but, the developing experience is really important, for example:

__Why ethereum got such many developers and users?__

Except the business operating and the smart-contract concepts, even publishing coin functions, one of the reasons on the core thing, the programs, they do have great programs, easy to use, easy to downloads.

Developers are happy to use it,  the developing process endues them senses of accomplishment, the hope of the future, they talking about it, writing articles about it, publishing programs about it —— more and more people join.

Every CITA show in the public view from company usage,  it is reach to the Ethereum, there are big needs and markets in China, the Chinese people, who expecting the changing of the time, who enjoy new technologies, who are dreaming about the future the whole lives, somebody has faith in blockchain, they do need to believe, something is happening.

emmmm, anyway, I think it is important, but I am not sure if it is the most, maybe the `different OS` testing problem can be delayed to `cita-1.0.0-stable`? 

It is quite nice to add: 

+ brew
+ snap

or

+ Linux-basic-version
+ MacOS-basic-version
+ Windows-basic-version

in the current downloads list.